pkg/build/strategies/sti/sti.go

```diff
@@ -642,7 +642,7 @@ func (builder *STI) Execute(command string, user string, config *api.Config) err
 	}
 	// Do not wait for source input if the container times out.
 	// FIXME: this potentially leaks a goroutine.
-	if !util.IsTimeoutError(err) {
+	if err != nil {
```

Comment:
*Disclaimer: I'd need some good time to re-understand all the little nuances here.*

---

Reading this feels really strange "if RunContainer returns a non-nil error, then wait"?!

Where "wait" means "if config.LayeredBuild is false, wait until 'source injection' is complete"...

IIRC the original problem is that 'source injection' misbehaves -- maybe it tended to hang, or maybe we proceeded too early before it was finished. To what I can understand there's a snow-ball effect going on here.

@csrwng @gabemontero are you able to reason about the flow here in a cleaner way?